docs/training/apis-howto/api-pytorch-trainer-ug.rst

```diff
@@ -0,0 +1,354 @@
+   +             test_mode=True
+             )
+
+This is the same codepath as
+
+.. code:: bash
+
+   det e create det.yaml . --local --test
+
+**************************************************************************
+ Step 4: Prepare Your Training Code for Deploying to a Determined Cluster
+**************************************************************************
+
+Once you are satisfied with the results of training the model locally, you submit the code to a
+cluster.
+
+**Example workflow of frequent iterations between local debugging and cluster deployment**
+
+This code should allow for local and cluster training with no code changes.
+
+.. code::
+
+    def main():
+   +   local = det.get_cluster_info() is None
+   +   if local:
+   +       # (Optional) Initialize distributed backend before pytorch.init()
+   +       dist.init_process_group(backend="gloo|nccl")
+   +       # Set flag used by internal PyTorch training loop
+   +       os.environ["USE_TORCH_DISTRIBUTED"] = "true"
+   +       distributed_context = core.DistributedContext.from_torch_distributed  (chief_ip="localhost")
+   +       # (Optional) Pass in an exp conf and instance of hparams if training code needs it
+   +       expconf = yaml.safe_load(pathlib.Path("./config.yaml"))
+   +       hparams = {"lr": 0.02}
+   +   else:
+   +       hparams = det.get_cluster_info().trial.hparams
+   +       expconf = None
+   +       distributed_context = None
+
+   +     with det.pytorch.init(
+   +       hparams=hparams,
+   +       exp_conf=expconf,
+   +       distributed=distributed_context
+         ) as train_context:
+             trial = MNistTrial(train_context)
+             trainer = det.pytorch.Trainer(trial, train_context)
+             trainer.fit(
+                 max_length=pytorch.Epoch(1),
+                 checkpoint_period=pytorch.Batch(10),
+                 validation_period=pytorch.Batch(10),
+             )
+
+**To run Trainer API solely on-cluster, the code is much simpler**
+
+.. code::
+
+   def on_cluster():
+       """
+       On-cluster training with Trainer API (entrypoint: python3 train.py)
+       """
+       hparams = det.get_cluster_info().trial.hparams
+
+       with det.pytorch.init() as train_context:
+           trial_inst = model.MNistTrial(train_context, hparams)
+           trainer = det.pytorch.Trainer(trial_inst, train_context)
+           trainer.fit(
+               max_length=pytorch.Epoch(1),
+               checkpoint_period=pytorch.Batch(10),
+               validation_period=pytorch.Batch(10),
+           )
+
+***************************************************
+ Step 5: Submit Your Trial for Training on Cluster
+***************************************************
+
+To run your experiment on cluster, you'll need to create a YAML file. Your YAML file must contain
+searcher configuration and entrypoint.
+
+.. note::
+
+   `global_batch_size` is required if `max_length` is configured in records
```

Comment:
recommend omitting; this is covered by expconf reference and has good error messages if you mess it up.